### ISSUE TITLE
UILayoutComponent refactor

### ISSUE THREAD

**Issue by Shulepov** [cocos2d/cocos2d-x#16131]
Currently UILayoutComponent contains booleans and separate values to represent type and value for position/size. Also it have separate values for margins for some reason.

It would be much simpler to leave only next values and calculate layout based on them:

```
HorizontalEdge _horizontalEdge;
PositionType _positionTypeX;
float _positionX;
SizeType _sizeTypeWidth;
float _width;

VerticalEdge _verticalEdge;
PositionType _positionTypeY;
float _positionY;
SizeType _sizeTypeHeight;
float _height;
```

Where:

```
enum class PositionType {
    ABSOLUTE,
    PERCENT
}

enum class SizeType {
     ABSOLUTE,
     PERCENT,
     RELATIVE_TO_PARENT
}
```

And that will cover all possible situations much simpler, including fixed margins that currently is not possible. 

Though to write it correctly it will require to break backward compatibility...

**Comment by gestern** [comment]
In this case (without values aggregates) it would take more CPU time to calculate interim values and have to break backward compatibility, am i right?

**Comment by Shulepov** [comment]
Currently there are also no values cache and position/size recalculated each time in `refreshLayout`. But it recalculated based on margins, which also calculated frequently. I think calculating margins isn't really necessary. So with my approach it may be even more performant.

But it really will break backward compatibility, that is a big issue.

I have read current sources and really can't understand how some parameters effect on the result. Also looks like sometimes order in which functions called is important. May be things not so bad as i think, but currently it's really quite hard to follow this code.

**Comment by gestern** [comment]
Will need research, while these values are used also by 3rd party components like Cocos Studio, JS etc.

**Comment by minggo** [comment]
@Shulepov break compatibility is not acceptable in v3, it may be considered in v4.
